janusgraph-core/src/main/java/org/janusgraph/graphdb/util/MultiIterator.java

```diff
@@ -0,0 +1,52 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.util;
+
+import org.apache.tinkerpop.gremlin.structure.util.CloseableIterator;
+
+import java.util.List;
+import java.util.Objects;
+
+public class MultiIterator<E> extends CloseableAbstractIterator<E> {
+
+    private List<CloseableIterator<E>> iterators;
+    private int current;
+
+    public MultiIterator(final List<CloseableIterator<E>> iterators) {
+        Objects.requireNonNull(iterators);
+        this.iterators = iterators;
+    }
+
+    @Override
+    protected E computeNext() {
+        while (current < iterators.size()) {
+            CloseableIterator<E> iterator = iterators.get(current);
```

Comment:
That would be a bit more complicated because it requires two instance variables: a current iterator and an iterator of iterators. May I know why you think one should avoid calling `get(index)` on `List`?

Reply:
I guess that's because `List` isn't guaranteed to be `ArrayList` which supports Random Access in `O(1)` time. Thus, someone may for some reason pass `LinkedList` to the constructor of this class. If so, `get` operation will cost `O(n)` time instead of `O(1)` time. Making this loop to be `O(n^2)` time. That said, it depends on `List` implementation which is used here

Reply:
I think that using a current iterator and an iterator of iterator is a more safer approach here

Reply:
Thanks @porunov, that's exactly what I wanted to point out

Reply:
However, I would prefer not to enforce `ArrayList`. Using an iterator over an arbitrary list should be equally es fast but maintains more flexibility. Especially in case we decide to switch back to `List` for some reason in the future.

Reply:
@rngcntr Agree with your points. Using iterator will have more sense here.